pkg/cli/preview/recorder.go

```diff
@@ -0,0 +1,266 @@
+	// eventType is the type of event
+	eventType EventType
+	// diff is the diff that was recorded
+	diff *structuredreporting.Diff
+	// kubeAction is the kube action that was recorded
+	kubeAction *kubeAction
+	// gcpAction is the gcp action that was recorded
+	gcpAction *gcpAction
+	// object is the object that was reconciled
+	object *unstructured.Unstructured
+}
+
+type EventType string
+
+const (
+	EventTypeReconcileStart EventType = "reconcileStart"
+	EventTypeDiff           EventType = "diff"
+	EventTypeKubeAction     EventType = "kubeAction"
+	EventTypeGCPAction      EventType = "gcpAction"
+)
+
+// kubeAction holds a kubernetes action that was recorded
+type kubeAction struct {
+	method string
+	action Action
+}
+
+// Action indicates whether we blocked or ignored the requested action.
+type Action string
+
+const (
+	// ActionIgnored indicates that the action was ignored.
+	ActionIgnored Action = "ignored"
+	// ActionBlocked indicates that the action was blocked.
+	ActionBlocked Action = "blocked"
+)
+
+// gcpAction holds a GCP action that was recorded
+type gcpAction struct {
+	method string
+	url    string
+	body   string
+	action Action
+}
+
+// NewStructuredReportingListener creates a new StructuredReportingListener.
+func (r *Recorder) NewStructuredReportingListener() structuredreporting.Listener {
+	return &structuredReportingListener{recorder: r}
+}
+
+// structuredReportingListener is a listener for structured reporting events.
+type structuredReportingListener struct {
+	recorder *Recorder
+}
+
+// OnError is called by the structured reporting subsystem when an error occurs.
+func (l *structuredReportingListener) OnError(ctx context.Context, err error, args ...any) {
+	blockedGCPError, ok := ExtractBlockedGCPError(err)
+	if !ok {
+		return
+	}
+	l.recorder.recordGCPAction(ctx, blockedGCPError, args, ActionBlocked)
+}
+
+// OnReconcileStart is called by the structured reporting subsystem when a reconcile starts.
+func (l *structuredReportingListener) OnReconcileStart(ctx context.Context, u *unstructured.Unstructured) {
+	l.recorder.recordReconcileStart(ctx, u)
+}
+
+// OnDiff is called by the structured reporting subsystem when a diff occurs.
+func (l *structuredReportingListener) OnDiff(ctx context.Context, diff *structuredreporting.Diff) {
+	l.recorder.recordDiff(ctx, diff)
+}
+
+// RecordBlockedKubeMethod is called by the interceptingKubeClient when a write operation is blocked.
+func (r *Recorder) RecordBlockedKubeMethod(ctx context.Context, method string, args ...any) {
+	r.recordKubeAction(ctx, method, args, ActionBlocked)
+}
+
+// RecordIgnoredKubeMethod is called by the interceptingKubeClient when a read operation is ignored.
+func (r *Recorder) RecordIgnoredKubeMethod(ctx context.Context, method string, args ...any) {
+	r.recordKubeAction(ctx, method, args, ActionIgnored)
+}
+
+// recordDiff captures the diff into our recorder.
+func (r *Recorder) recordDiff(ctx context.Context, diff *structuredreporting.Diff) {
```

Comment:
Slight concern this is being done on the primary goroutine of the reconcile. My concern is what happens if we have large volumes of data and a slow disk. Also what happens if the disk runs out of room?

Reply:
So even in "preview mode" it only goes to memory.  In existing modes of operation we expect no listeners.

I will add a warning to the interface though, saying you should treat this as a logger and generally copy the data and queue it for another thread if you are going to do anything substantial with it (like writing to disk)